libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,584 @@
+        } else {
+            for (int16_t i = start_index + 1; i < end_index; i++) {
+                _simplify_bitmask.clear(i);
+            }
+        }
+    }
+    _simplify_complete = true;
+}
+
+/**
+*   This method runs for the allotted time, and detects loops in a path. All detected loops are added to _prunable_loops,
+*   this function does not alter the path in memory. It works by comparing the line segment between any two sequential points
+*   to the line segment between any other two sequential points. If they get close enough, anything between them could be pruned.
+*
+*   Note that this method might take a bit longer than LOOP_TIME. It only stops after it's already run longer.
+*/
+void AP_SafeRTL::detect_loops()
+{
+    // if SafeRTL is not active OR if this algorithm has already run to completion OR there's fewer than 3 points in the path
+    if (!_active || _prune_complete || _last_index < 3) {
+        return;
+    }
+    uint32_t start_time = AP_HAL::micros();
+
+    while (_prune_current_i < _last_index - 1) {
+        // if this method has run for long enough, exit
+        if (AP_HAL::micros() - start_time > SAFERTL_LOOP_TIME_US) {
+            return;
+        }
+
+        // this check prevents detection of a loop-within-a-loop
+        int16_t j = MAX(_prune_current_i + 2, _prune_min_j);
+        while (j < _last_index) {
+            dist_point dp = segment_segment_dist(_path[_prune_current_i], _path[_prune_current_i+1], _path[j], _path[j+1]);
+            if (dp.distance <= SAFERTL_PRUNING_DELTA) { // if there is a loop here
+                _prune_min_j = j;
+                // if the buffer is full
+                if ( _prunable_loops_last_index >= _path_points_max * SAFERTL_LOOP_BUFFER_LEN_MULT - 1) {
+                    _prune_complete = true; // pruning is effectively complete now, since there's no reason to continue looking for them.
+                    return;
+                }
+                // int promotion rules disallow using i+1 and j+1
+                _prunable_loops[++_prunable_loops_last_index] = {(++_prune_current_i)--,(++j)--,dp.point};
+            }
+            j++;
+        }
+        _prune_current_i++;
+    }
+    _prune_complete = true;
+}
+
+//
+// Private methods
+//
+
+/**
+*   Run this regularly, in the main loop (don't worry - it runs quickly, 100us). If no cleanup is needed, it will immediately return.
+*   Otherwise, it will run a cleanup, based on info computed by the background methods, detect_simplifications() and detect_loops().
+*   If no cleanup is possible, this method returns false. This should be treated as an error condition.
+*/
+bool AP_SafeRTL::routine_cleanup()
+{
+    // We only do a routine cleanup if the memory is almost full. Cleanup deletes
+    // points which are potentially useful, so it would be bad to clean up if we don't have to
+    if (_last_index < _path_points_max - 10) {
```

Comment:
I'm not sure we really want to defer this till the buffers almost full. It has the effect of causing the system to have to spend more time (if it simplifies the first points/anything early in the loop), and increases the amount of time it takes to make it through the simplification/cleanup algorithms. I'd suggest we should always take a stab at simplifying/delooping if there are say 20 new points since we last tried?